src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioPortableExecutableReference.FileTimeStampProvider.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+
+internal partial class VisualStudioMetadataReferenceManager
+{
+    private sealed partial class VisualStudioPortableExecutableReference
+    {
+        private class FileTimeStampProvider
+        {
+            private readonly Dictionary<string, DateTime> _cachedTimestamps = [];
+            private readonly Dictionary<string, IFileChangeContext> _fileChangeContexts = [];
+
+            public DateTime GetTimeStamp(string fullPath, IFileChangeWatcher watcher)
+            {
+                lock (_cachedTimestamps)
+                {
+                    // Attempt to use the cached timestamp for this file. If we don't have a cached timestamp,
+                    // we'll need to recalculate
+                    if (!_cachedTimestamps.TryGetValue(fullPath, out var timestamp))
+                        return timestamp;
+
+                    var directory = Path.GetDirectoryName(fullPath);
+                    var context = _fileChangeContexts.GetOrAdd(directory, static (directory, arg) =>
+                    {
+                        var (watcher, self) = arg;
+                        var context = watcher.CreateContext([new WatchedDirectory(directory, extensionFilters: [])]);
+                        context.FileChanged += self.OnContextFileChanged;
+
+                        return context;
+                    }, (watcher, this));
```

Comment:
how expensive is this?  should we consider doing outside of lock?

Reply:
This is cheap (at least for the FileChangeWatcher.CreateContext implementation)

Reply:
Unless you were talking about the GetFileTimeStamp call, which can be expensive. I could be convinced to do that outside the lock and then reobtain the lock.

Reply:
Went ahead and moved the actual GetFileTimeStamp call outside the lock